lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java

```
@@ -0,0 +1,194 @@
+    throw new IllegalStateException("Monitor is readOnly cannot delete queries");
+  }
+
+  @Override
+  public void clear() throws IOException {
+    throw new IllegalStateException("Monitor is readOnly cannot clear");
+  }
+
+  @Override
+  public long getLastPurged() {
+    return -1;
+  }
+
+  @Override
+  public void addListener(MonitorUpdateListener listener) {
+    throw new IllegalStateException("Monitor is readOnly cannot register listeners");
+  }
+
+  // ---------------------------------------------
+  //  Helper classes...
+  // ---------------------------------------------
+
+  /** A Collector that decodes the stored query for each document hit reparsing them everytime. */
+  static final class LazyMonitorQueryCollector extends SimpleCollector {
+    private final QueryIndex.QueryCollector matcher;
+    private final QueryIndex.DataValues dataValues = new QueryIndex.DataValues();
+    private final MonitorQuerySerializer serializer;
+    private final QueryDecomposer decomposer;
+
+    LazyMonitorQueryCollector(
+        QueryIndex.QueryCollector matcher,
+        MonitorQuerySerializer serializer,
+        QueryDecomposer decomposer) {
+      this.matcher = matcher;
+      this.serializer = serializer;
+      this.decomposer = decomposer;
+    }
+
+    @Override
+    public void setScorer(Scorable scorer) {
+      this.dataValues.scorer = scorer;
+    }
+
+    @Override
+    public void collect(int doc) throws IOException {
+      dataValues.advanceTo(doc);
+      BytesRef cache_id = dataValues.cacheId.lookupOrd(dataValues.cacheId.ordValue());
+      BytesRef query_id = dataValues.queryId.lookupOrd(dataValues.queryId.ordValue());
+      MonitorQuery mq = serializer.deserialize(dataValues.mq.binaryValue());
+      QueryCacheEntry query =
+          QueryCacheEntry.decompose(mq, decomposer).stream()
+              .filter(queryCacheEntry -> queryCacheEntry.cacheId.equals(cache_id.utf8ToString()))
+              .findFirst()
+              .orElse(null);
```

Can we throw an exception here instead? `null` is going to cause problems in the collector I think?